Improve AI squad pathing and regrouping behavior.

RoosterDragon opened the pull request:
Depends on #20227. Fixes #17327 and https://github.com/OpenRA/OpenRA/issues/17327#issuecomment-554163061. Fixes #3763.

----

Ensure the target location can be pathed to by all units in the squad, so the squad won't get stuck if some units can't make it. Improve the choice of leader for the squad. We attempt to a choose a leader whose locomotor is the most restrictive in terms of passable terrain. This maximises the chance that the squad will be able to follow the leader along the path to the target. ~~We also choose a leader based on the remaining movement length to the target rather than by physical distance. This avoids the squad regrouping backwards and getting stuck if the squad has to move away from the target to go around an obstacle.~~ We also keep this choice of leader as the squad advances, this avoids the squad constantly switching leaders and regrouping backwards in some cases.

dnqbob commented:
I think I have a more performance friendly (maybe) and simple idea:

(1). Leader won't change until its death. Therefore, we don't have to calculate group center every time.
(2). All other units only regroup to leader instead of doing a long path finding. This should save some performances considering target is always far away for a bunch of units to path find.

There are some shortcomings for (2) on my own application when at tunnel and overpass bridge, which may need some techniques on making leader to see if the squad is stuck.

There is some [code](https://github.com/MustaphaTR/OpenRA/blob/rv-engine/OpenRA.Mods.Common/Traits/BotModules/Squads/States/GroundStates.cs) related.

RoosterDragon committed:
Improve AI squad pathing and regrouping behavior.

Ensure the target location can be pathed to by all units in the squad, so the squad won't get stuck if some units can't make it. Improve the choice of leader for the squad. We attempt to a choose a leader whose locomotor is the most restrictive in terms of passable terrain. This maximises the chance that the squad will be able to follow the leader along the path to the target. We also keep this choice of leader as the squad advances, this avoids the squad constantly switching leaders and regrouping backwards in some cases.

RoosterDragon commented:
@dnqbob I implemented your first suggestion to keep the leader. This is nice as we don't have to dig into the remaining length in Move activites any more either which simplifies this change. Since you mentioned issues with the second item I will leave that out for now.

I have checked that this still resolves #17327 on the provided test case map.

PunkPun commented:
needs a rebase

RoosterDragon commented:
Rebased

PunkPun reviewed (commented):
[MapAndReplays.zip](https://github.com/OpenRA/OpenRA/files/12557516/MapAndReplays.zip)


The Ai's squads still seem to be acting very incorrectly. See submarines and cruisers in this replay

PunkPun commented:
This fixes https://github.com/OpenRA/OpenRA/issues/17327#issuecomment-554163061, but I can't see any fixes to any of the other aforementioned issues, including #21046

Fixing #21046 will likely be quite a bit of scope creep is I'm fine with merging this now

PunkPun commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/b5857aaeb4317a113efa85a1011d7a97147432a4...daf3a4b2eba5e151db66225cb05fd3104b035a3f)
